### ISSUE TITLE
Have file globbing only match files that are checked into a source repository (git/hg/etc.)

### ISSUE THREAD

**Issue by tfar** [coala-analyzer/coala#2026]
It would be great if file globbing would only match files that are checked into a source code repository. So simply calling a git command to get a list of files and then have globbing be applied to that list. This would automatically integrate with .gitignore.

Additionally a flag to only look at changed files (the one that git/hg/etc. think are changed) would be great because that way, you could run in before you commit and don't have expensive checks look at the whole project.

**Comment by Makman2** [comment]
Only-for-changed-files feature is currently coming: https://github.com/coala-analyzer/coala/pull/2016 :)
For the git-globbing: I like that idea. How do you want to handle untracked files that are not ignored?

**Comment by sils1297** [comment]
We'd probably want a very light integration, i.e. just an option "apply globs only to files tracked by Git" and then implement support in our globs for that?

CC @fneu

**Comment by fneu** [comment]
I agree with @sils1297. Let's add an option to only check on files that are included in the current Version control system (only git).

This does not include untracked files.

Now we could either run globs normally and filter out the ones not in git or we could fnmatch all files reporter by git.

Usually, glob is orders of magnitude faster (fnmatch checks on all files, glob checks less and less files as it progresses along the pattern) but there are corner cases like checking for `'**/*'` inside a big repository.

**Comment by tfar** [comment]
Haven't used it but I just stumbled upon http://anyvcs.readthedocs.org/en/latest/anyvcs.html. It's a python lib for providing directory listings or changed file listings of git/hg/svn repositories.

**Comment by sils1297** [comment]
Looks neat, there's a huge problem with this one though: it isn't maintained. Last commit was in 2014, there are issues as well as pull requests unanswered, the author seems to be totally gone so we'd have to maintain a fork more or less. I'm pretty sure there are alternatives though, using an abstraction is definitely the right way to go (that's also why I wrote [IGitt](https://gitlab.com/gitmate/IGitt) before I started interfacing GitHub for [GitMate](http://gitmate.io/))

**Comment by tfar** [comment]
Oh, didn't notice that. However, it seems most python libraries for abstracting version control systems seem to be unmaintained.

**Comment by sils1297** [comment]
Yeah we'll only need a very small subset of commands, possibly a very simple function does the job better and sleeker than a lib

**Comment by sils** [comment]
Hi, this hasn't been picked up lately but here's an update on what I think about this:

From usability testing people expect from coala that it obeys the gitignore, they would not expect it to only check files that are in VCS.

We already have an algorithm in coala-quickstart by @adtac that parses the gitignore and makes coala ignore globs out of it - maybe we can use that and create a cached preprocessed file which we update if we detect that the gitignore has been modified similar to our already existing results caching algorithm (again written by @adtac ... spotting a pattern here). This way we don't loose too much performance by parsing/translating globs, we can reuse most of the code that we already have and only tie it together neatly. It also seems to be what users expect and I don't see people being surprised if that is the default behaviour, if there's a gitignore we can log that we use it in the debug channel.

I think @underyx was for this sort of idea as well so CCing him here.

**Comment by adtac** [comment]
Probably obvious, but I just want to explicitly that this cache can be stored in the same project cache file (see `~/.local/share/coala`) that we already store.

**Comment by Siddhartha1234** [comment]
@sils Should tracking of `.gitignore`  be added to `coala-quickstart` to  update `ignored_globs` in the settings or should this behavior be handled by default in `coala` and removed from `coala-quickstart` completely ? There should be more discussion regarding which features of VCS integration should be allowed by default and which others optional features can to be added in it IMO.
 CC @adtac .

**Comment by adtac** [comment]
Neither. This issue wants to automatic gitignore tracking (that is, instead of giving `ignore = `, I'd give `automatic_ignore = true` or something like that). So coala-quickstart will be untouched.

**Comment by sils** [comment]
actually we can drop the gitignore handling from coala-quickstart if this
is implemented - less duplication. And I'm not sure I'd do an
`automatic_ignore` rather set the gitignore values by default for any
section maybe? If the user doesn't want it he can still override the
ignore's if he wants it he can +=
